### PR TITLE
[FIX] (PERF) Running an inventory adjustment on a product using lot tracking

### PR DESCRIPTION
**Pull request by xdu-odoo**
Description of the issue/feature this PR addresses:
Running an inventory adjustment on a product using lot tracking and having a lot of lots could take more than 1 hour because of slow create method in inventory line class. This commit manage to reduce that time to 3 or 4 minutes on the same use case.

Current behavior before PR:
All computed fields are recomputed at each insert/update which is very slow.

Desired behavior after PR is merged:
This commit rewrites the inventory with pure SQL :
- insert directly without using ORM and thus preventing useless queries
- marks all fields for all the inserted lines as modified at once
- trigger recompute at the end

opw 1959459

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/stock/models/stock_inventory.py [hunk 1]
```diff
@@ -200,13 +200,54 @@ def action_cancel_draft(self):
...
+                lines.modified(self.env['stock.inventory.line']._fields)
+
+            vals = {'state': 'confirm', 'date': fields.Datetime.now()}
             inventory.write(vals)
+        self.recompute()
         return True
+
     prepare_inventory = action_start
 
+    def inventory_line_to_sql_insert(self, iline, inventory):
```
Comment (fmdl) [inline addons/stock/models/stock_inventory.py]: could put this methode private ?
add check_acces_right() to check right to make the SQL query.

cc @odony

**Comment by fmdl** [comment]
@xmo-odoo are you try with `with env.norecompute` ?

**Comment by xdu-odoo** [comment]
> @xmo-odoo are you try with `with env.norecompute` ?

As @rco-odoo told me, when writing in a one2many field the norecompute env is automatically enabled and indeed the tests I ran showed no benefit

**Comment by fmdl** [comment]
if not, use `create` fonction indeed `write + (0, 0, vals)`

```python
with env.norecompute():
    for v in values:
         self.env['xx.xx'].create(v)
```

**Comment by fmdl** [comment]
Please find some idea to improve perf (test make in V11). It is more than issue with cache and related field `type`, `uom_id` of `product.product`.
With this PR https://github.com/odoo/odoo/pull/32748, before 30s, after 17s with 1000 products. ==> + 76%
The times to `create()` the 1000 stock.inventoty.line, without recompute is now 7 secondes (before 17 secondes).
Issues:
1: this line is horrible because uom_id is computed for each product (it is a computed field), and not for all product in one time. Note in my PR I take the uom_id direct in the SQL query
2: each time you create a new inventory.line `_check_product_id()` https://github.com/odoo/odoo/blob/11.0/addons/stock/models/stock_inventory.py#L403 is called. But the product.type is not in cache. that why I have add some `read` to put in cache the type before the create (this win 10 secondes).
3: eact time you `create` the `_compute_theoretical_qty()` https://github.com/odoo/odoo/blob/11.0/addons/stock/models/stock_inventory.py#L352 is called, with `norecompute` is called one time.


Note : to put in cache 1000 `uom_id` and `type` of `product.product` one by one is 10 secondes. In one time is 0.46 secondes. :-)

cc @odony

File: addons/stock/models/stock_inventory.py [hunk 2]
```diff
@@ -200,13 +200,48 @@ def action_cancel_draft(self):
...
+        return self.env.cr.fetchone()[0]
+
+    def _inventory_line_values_to_row(self, values):
+        return {
+            "inventory_id": self.id,
+            "product_id": values['product_id'] or None,
+            "product_uom_id": values.get('product_uom_id') or None,
+            "prod_lot_id": values.get('prod_lot_id') or None,
+            "package_id": values.get('package_id') or None,
+            "theoretical_qty": values.get('theoretical_qty', 0),
```
Comment (fmdl) [inline addons/stock/models/stock_inventory.py]: `theoretical_qty` is a computed field
@rco-odoo

File: addons/stock/models/stock_inventory.py [hunk 3]
```diff
@@ -200,13 +200,48 @@ def action_cancel_draft(self):
...
+
+            vals = {'state': 'confirm', 'date': fields.Datetime.now()}
             inventory.write(vals)
+        self.recompute()
         return True
+
     prepare_inventory = action_start
 
+    def _insert_inventory_line(self, values):
+        row = self._inventory_line_values_to_row(values)
```
Comment (fmdl) [inline addons/stock/models/stock_inventory.py]: if you have access to `write` on stock.inventory, but no access to `create` stock.inventoty.line, with this SQL methode you can do that.
maybe add `self.env['stock.inventory.line].check_access_right('create')`.

File: addons/stock/models/stock_inventory.py [hunk 4]
```diff
@@ -200,13 +200,48 @@ def action_cancel_draft(self):
...
+        params = (self.env.uid, self.env.uid) + tuple(row.values())
+        query = 'INSERT INTO "stock_inventory_line" ({}) VALUES ({}) RETURNING id'
+        self.env.cr.execute(
+            query.format(", ".join('"%s"' % col for col in cols), ", ".join(fmts)),
+            params,
+        )
+        return self.env.cr.fetchone()[0]
+
+    def _inventory_line_values_to_row(self, values):
+        return {
```
Comment (fmdl) [inline addons/stock/models/stock_inventory.py]: add `self.ensure_one()` no ?

File: addons/stock/models/stock_inventory.py (hunk 3 above)
Comment (rco-odoo) [reply addons/stock/models/stock_inventory.py]: @fmdl strictly speaking, you are right. But having `write` access on `stock.inventory` and no access to create `stock.inventory.line` simply does not make sense business-wise. As the method is not public, I don't see the point into enforcing such a security policy.

File: addons/stock/models/stock_inventory.py (hunk 4 above)
Comment (rco-odoo) [reply addons/stock/models/stock_inventory.py]: @fmdl come on...

File: addons/stock/models/stock_inventory.py (hunk 2 above)
Comment (rco-odoo) [reply addons/stock/models/stock_inventory.py]: @fmdl you are right. This line is pointless.

**Comment by fmdl** [comment]
@rco-odoo, question in V12 with `create_multi `, this PR should be ported in V12 ?
Have you see other perf issue (see comment below) ?

**Pull request closed without merging**